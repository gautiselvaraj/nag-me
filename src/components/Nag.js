import React, {Component} from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import humanizeDuration from 'humanize-duration';
import {StaggeredMotion, spring} from 'react-motion';
import {humanizeNextNag} from '../utils/time';
import Button from './Button';
import Icon from './Icon';

const NagWrap = styled.div`
  background-color: #ffffff;
  border: 2px solid transparent;
  border-radius: 4px;
  box-shadow: 0 0 2px rgba(0, 0, 0, .25);
  padding: 5px;
  will-change: transform;
`;

const NagHeading = styled.h3`
  font-size: 1.15rem;
  margin: 0;
`;

const NagTimer = styled.p`
  color: ${props => props.theme.greyDarker};
  font-size: .85rem;
  margin-bottom: 0;
  margin-top: 5px;
`;

const NagRepeat = styled.p`
  color: ${props => props.theme.greyLight};
  font-size: .75rem;
  margin-bottom: 0;
  margin-top: 5px;
`;

const NagEditLink = styled.div`
  position: absolute;
  right: -45px;
  top: 50%;
  will-change: transform;
`;

const NagDeleteLink = styled.div`
  position: absolute;
  right: -85px;
  top: 50%;
  will-change: transform;
`;

const NagStateLink = styled.div`
  position: absolute;
  right: -125px;
  top: 50%;
  will-change: transform;
`;

const nagSpring = {stiffness: 250, damping: 22};

export default class Nag extends Component {
  static propTypes = {
    nag: PropTypes.object.isRequired
  };

  state = {nowTimestamp: Date.now(), active: false};

  updateTimestamp = () => {
    this.setState({nowTimestamp: Date.now()});
  }

  getInterpolatedStyles = prevInterpolatedStyles => {
    const endValue = prevInterpolatedStyles.map((_, i) => {
      if(this.state.active) {
        return i === 0 ?
          {t: spring(-135, nagSpring)} :
          {t: spring(prevInterpolatedStyles[i - 1].t, nagSpring)};
      } else {
        return i === 3 ?
          {t: spring(0, nagSpring)} :
          {t: spring(prevInterpolatedStyles[i + 1].t, nagSpring)};
        }
    });
    return endValue;
  };

  handleClick = () => {
    this.setState({active: !this.state.active})
  }

  componentDidMount() {
    this.intervalTimer = setInterval(this.updateTimestamp, 1000);
  }

  componentDidUnMount() {
    clearInterval(this.intervalTimer);
  }

  render() {
    const {nag} = this.props;
    const {nowTimestamp} = this.state;

    return (
      <StaggeredMotion
        defaultStyles={Array.from(Array(4)).map(i => ({t: 0}))}
        styles={this.getInterpolatedStyles}>
        {interpolatingStyles =>
          <div onClick={this.handleClick}>
            {interpolatingStyles.map(({t}, i) => (
              <div key={i}>
                {i === 0 &&
                  <NagWrap style={{
                    WebkitTransform: `translateX(${t}px)`,
                    transform: `translateX(${t}px)`
                  }}>
                    <NagHeading>{nag.title}</NagHeading>
                    <NagTimer>in {humanizeNextNag(nag.nextNag - nowTimestamp)}</NagTimer>
                    <NagRepeat>
                      {!!nag.repeats &&
                        <span>and repeat every {humanizeDuration(nag.repeats)}</span>
                      }
                    </NagRepeat>
                  </NagWrap>
                }
                {i === 1 &&
                  <NagEditLink style={{
                    WebkitTransform: `translate(${t}px, -50%)`,
                    transform: `translate(${t}px, -50%)`
                  }}>
                    <Button circle title={`Edit ${nag.title} Nag`}>
                      <Icon edit />
                    </Button>
                  </NagEditLink>
                }
                {i === 2 &&
                  <NagDeleteLink style={{
                    WebkitTransform: `translate(${t}px, -50%)`,
                    transform: `translate(${t}px, -50%)`
                  }}>
                    <Button circle title={`Delete ${nag.title} Nag`}>
                      <Icon delete />
                    </Button>
                  </NagDeleteLink>
                }
                {i === 3 &&
                  <NagStateLink style={{
                    WebkitTransform: `translate(${t}px, -50%)`,
                    transform: `translate(${t}px, -50%)`
                  }}>
                    {nag.paused ?
                      <Button circle title={`Resume ${nag.title} Nag`}>
                        <Icon resume />
                      </Button> :
                      <Button circle title={`Pause ${nag.title} Nag`}>
                        <Icon pause />
                      </Button>
                    }
                  </NagStateLink>
                }
              </div>
            ))}
          </div>
        }
      </StaggeredMotion>
    )
  }
}
