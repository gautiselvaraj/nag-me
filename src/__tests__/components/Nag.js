import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import Nag, {
  NagTimer,
  NagRepeat,
  NagStatus,
  NagProgress,
  NagEditLink,
  NagDeleteLink,
  NagStateLink
} from '../../components/Nag';
import 'jest-styled-components';

describe('<Nag />', () => {
  let nagEdit,
    nagPause,
    nagResume,
    nagDelete,
    nagStatusUpdate,
    nagObject,
    dateNow;

  beforeAll(() => {
    dateNow = window.Date.now;
    window.Date.now = jest.fn(() => 1517227842390);
  });

  afterAll(() => {
    window.Date.now = dateNow;
  });

  beforeEach(() => {
    nagEdit = jest.fn();
    nagPause = jest.fn();
    nagResume = jest.fn();
    nagDelete = jest.fn();
    nagStatusUpdate = jest.fn();
    nagObject = {
      id: 1,
      title: 'Test Nag 1',
      firstNag: 1234567890,
      nextNag: 1234567890,
      repeats: '1 min',
      status: 'LIVE',
      createdAt: 1234567890
    };
  });

  it('should match snapshot', () => {
    const tree = renderer
      .create(
        <Nag
          nag={nagObject}
          nagEdit={nagEdit}
          nagPause={nagPause}
          nagResume={nagResume}
          nagDelete={nagDelete}
          nagStatusUpdate={nagStatusUpdate}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should have correct elements and children', () => {
    const nag = mount(
      <Nag
        nag={nagObject}
        nagEdit={nagEdit}
        nagPause={nagPause}
        nagResume={nagResume}
        nagDelete={nagDelete}
        nagStatusUpdate={nagStatusUpdate}
      />
    );
    expect(nag.find('h3').length).toBe(1);
    expect(nag.find('h3').text()).toBe('Test Nag 1');
    expect(nag.find(NagTimer).length).toBe(1);
    expect(nag.find(NagTimer).text()).toBe('in 48 years');
    expect(nag.find(NagRepeat).length).toBe(1);
    expect(nag.find(NagRepeat).text()).toBe('and repeat every 1 minute');
    expect(nag.find(NagStatus).length).toBe(0);
    expect(nag.find(NagProgress).length).toBe(1);
    expect(nag.find(NagEditLink).length).toBe(1);
    expect(nag.find(NagDeleteLink).length).toBe(1);
    expect(nag.find(NagStateLink).find('button[title="Pause"]').length).toBe(1);
  });

  it('should call nagEdit', () => {
    const nag = mount(
      <Nag
        nag={nagObject}
        nagEdit={nagEdit}
        nagPause={nagPause}
        nagResume={nagResume}
        nagDelete={nagDelete}
        nagStatusUpdate={nagStatusUpdate}
      />
    );
    nag
      .find(NagEditLink)
      .find('button')
      .simulate('click');
    expect(nagEdit.mock.calls.length).toBe(1);
  });

  it('should call nagPause', () => {
    const nag = mount(
      <Nag
        nag={nagObject}
        nagEdit={nagEdit}
        nagPause={nagPause}
        nagResume={nagResume}
        nagDelete={nagDelete}
        nagStatusUpdate={nagStatusUpdate}
      />
    );
    nag
      .find(NagStateLink)
      .find('button')
      .simulate('click');
    expect(nagPause.mock.calls.length).toBe(1);
  });

  it('should call nagDelete', () => {
    const nag = mount(
      <Nag
        nag={nagObject}
        nagEdit={nagEdit}
        nagPause={nagPause}
        nagResume={nagResume}
        nagDelete={nagDelete}
        nagStatusUpdate={nagStatusUpdate}
      />
    );
    nag
      .find(NagDeleteLink)
      .find('button')
      .simulate('click');
    expect(nagDelete.mock.calls.length).toBe(1);
  });

  it('should call nagResume if status is paused', () => {
    const nag = mount(
      <Nag
        nag={Object.assign({}, nagObject, { status: 'PAUSED' })}
        nagEdit={nagEdit}
        nagPause={nagPause}
        nagResume={nagResume}
        nagDelete={nagDelete}
        nagStatusUpdate={nagStatusUpdate}
      />
    );
    nag
      .find(NagStateLink)
      .find('button')
      .simulate('click');
    expect(nagResume.mock.calls.length).toBe(1);
  });
});
