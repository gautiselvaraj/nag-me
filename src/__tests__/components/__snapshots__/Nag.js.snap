// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nag /> should match snapshot 1`] = `
.c0 {
  background-color: #ffffff;
  border-radius: 4px;
  box-shadow: 0 0 2px rgba(0,0,0,0.25);
  padding: 10px;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  will-change: transform;
  border-bottom: 3px solid undefined;
}

.c1 {
  font-size: 1.2rem;
  margin: 0;
}

.c2 {
  font-size: 0.85rem;
  margin-bottom: 0;
  margin-top: 5px;
}

.c3 {
  font-size: 0.75rem;
  margin-bottom: 0;
  margin-top: 5px;
}

.c4 {
  border-top: 3px solid #6e8ea1;
  border-radius: 0 0 3px 3px;
  bottom: -3px;
  left: 0;
  position: absolute;
  -webkit-transition: width 1s;
  transition: width 1s;
  will-change: width;
}

.c5 {
  position: absolute;
  right: -45px;
  top: 50%;
  will-change: transform;
}

.c6 {
  border: 1px solid props => props.theme.mainDarken;
  cursor: pointer;
  display: inline-block;
  padding: 0;
  box-sizing: border-box;
  font-size: 0.85rem;
  line-height: 2rem;
  text-decoration: none;
  text-transform: uppercase;
  -webkit-transition-duration: 0.5s;
  transition-duration: 0.5s;
  -webkit-transition-property: background-color;
  transition-property: background-color;
  border-radius: 50%;
  height: 2rem;
  text-align: center;
  width: 2rem;
}

.c6 i {
  position: relative;
  top: 2px;
}

.c7 {
  font-family: icons;
  font-size: 1rem;
  font-style: normal;
  font-variant: normal;
  font-weight: normal;
  line-height: 1;
  speak: none;
  text-transform: none;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

.c7:before {
  content: "\\e906";
}

.c9 {
  font-family: icons;
  font-size: 1rem;
  font-style: normal;
  font-variant: normal;
  font-weight: normal;
  line-height: 1;
  speak: none;
  text-transform: none;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

.c9:before {
  content: "\\e9ac";
}

.c11 {
  font-family: icons;
  font-size: 1rem;
  font-style: normal;
  font-variant: normal;
  font-weight: normal;
  line-height: 1;
  speak: none;
  text-transform: none;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

.c11:before {
  content: "\\ea1d";
}

.c8 {
  position: absolute;
  right: -85px;
  top: 50%;
  will-change: transform;
}

.c10 {
  position: absolute;
  right: -125px;
  top: 50%;
  will-change: transform;
}

<div
  onClick={[Function]}
>
  <div>
    <div
      className="c0"
      style={
        Object {
          "WebkitTransform": "translateX(0px)",
          "transform": "translateX(0px)",
        }
      }
    >
      <h3
        className="c1"
      >
        Test Nag 1
      </h3>
      <div>
        <p
          className="c2"
        >
          in 
          48 years
        </p>
        <p
          className="c3"
        >
          <span>
            and repeat every 
            1 minute
          </span>
        </p>
      </div>
      <div
        className="c4"
        style={
          Object {
            "width": "2526655556.85%",
          }
        }
      />
    </div>
  </div>
  <div>
    <div
      className="c5"
      style={
        Object {
          "WebkitTransform": "translate(0px, -50%)",
          "transform": "translate(0px, -50%)",
        }
      }
    >
      <button
        className="c6"
        onClick={[Function]}
        title="Edit"
      >
        <i
          className="c7"
        />
      </button>
    </div>
  </div>
  <div>
    <div
      className="c8"
      style={
        Object {
          "WebkitTransform": "translate(0px, -50%)",
          "transform": "translate(0px, -50%)",
        }
      }
    >
      <button
        className="c6"
        onClick={[Function]}
        title="Delete"
      >
        <i
          className="c9"
        />
      </button>
    </div>
  </div>
  <div>
    <div
      className="c10"
      style={
        Object {
          "WebkitTransform": "translate(0px, -50%)",
          "transform": "translate(0px, -50%)",
        }
      }
    >
      <button
        className="c6"
        onClick={[Function]}
        title="Pause"
      >
        <i
          className="c11"
        />
      </button>
    </div>
  </div>
</div>
`;
